### ISSUE TITLE
Use macros for loop-unrolling

### ISSUE THREAD

**Issue by Rudxain** [BLAKE3-team/BLAKE3#269]
As I said [here](https://github.com/BLAKE3-team/BLAKE3/issues/7#issuecomment-1192180277), using macros has the benefit of performance without sacrificing maintainability.

I've seen that `reference_impl` (which is *supposed* to be simple and not optimized) [also has](https://github.com/BLAKE3-team/BLAKE3/blob/f84636e59ce575e5dd127399e0c7de0c1ea595da/reference_impl/reference_impl.rs#L102-L114) unrolled loops, for some reason. Could someone please explain why?

BTW, I'm willing to open a PR for this

**Comment by oconnor663** [comment]
In many cases I just find it easier to read and write. For example [these calls to the `g` function](https://github.com/BLAKE3-team/BLAKE3/blob/f84636e59ce575e5dd127399e0c7de0c1ea595da/reference_impl/reference_impl.rs#L56-L64) could be reduced to a couple loops, but those loops would be unnecessarily tricky. For the [`round` calls that you linked](https://github.com/BLAKE3-team/BLAKE3/blob/f84636e59ce575e5dd127399e0c7de0c1ea595da/reference_impl/reference_impl.rs#L102-L114), one nice thing about writing it out is that it highlights that there's no `permute` after the last round. If we were writing a loop, we could add an `if` statement to make the `permute` conditional (and assume that unrolling and constant propagation would delete the `if`), or we could just go ahead and do the extra `permute` (and assume the optimizer would notice that those writes aren't observed), or we could pull the first `round` out of the loop and switch it from `round`/`permute` to `permute`/`round`. But I like the way it is better than any of those options :)

**Comment by Rudxain** [comment]
> I just find it easier to read and write.

I think I get what you mean, it's easier to see a pattern in the iterations (sometimes)

> [these calls to the `g` function](https://github.com/BLAKE3-team/BLAKE3/blob/f84636e59ce575e5dd127399e0c7de0c1ea595da/reference_impl/reference_impl.rs#L56-L64) could be reduced to a couple loops.

Not necessarily. We could use an array of arrays, of length 8, where each array contains the args given to the fn on each iteration. The index to `m` can be an int that increments in steps of 2, or we could use `enumerate` and multiply the yielded index by 2 on each iteration.

> it highlights that there's no `permute` after the last round.

When I saw it, at first glance, I thought the number of `round`s was the same as `permute`s. Then I looked carefully and noticed the actual interweaving pattern layout. I would describe it as "the permutations are sandwiched between the rounds". I gotta admit I'm kinda hungry now, so that analogy was a product of a bias, lol.
 
> `if` statement to make the `permute` conditional

I prefer the following alternatives:

> do the extra `permute`

That would be confusing to other devs reading the code. If the last call isn't observable, then it would be misleading to make such a call.

> pull the first `round` out of the loop and switch it from `round`/`permute` to `permute`/`round`

Or pull out the last `round` and preserve the pattern (except for the unconditional call at the end)

> But I like the way it is better than any of those options :)

As I said earlier, I get why it's sometimes ok to do explicit unrolling, even in the ref-impl, **specially** for parts of the code that *shouldn't be touched* (like critical points in the algorithm), to reduce the likelihood of typos.

A typo in a rolled loop breaks all iterations. A typo in an unrolled loop breaks 1 iteration. But either way, the entire program becomes incorrect if there's just 1 typo, and unrolling requires more typing (therefore increasing the probability of a typo)

But I agree with you :) (I guess? I mean, I have my own preferences, but I respect yours)

Should I close this issue, or keep it open just-in-case anyone else wants to post some thoughts?
